### PR TITLE
H305: Physics BC enforcement — zero WSS normal component post-TTA

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The DrivAerML CFD solver enforces τ·n̂ = 0 exactly at every surface point — the no-penetration boundary condition means wall shear stress is purely tangential. Our neural network has **no such constraint at inference**: it can predict a non-zero normal component τ·n̂ ≠ 0. Subtracting this normal component post-hoc is a physics-guaranteed, **zero-cost correction** (zero extra forward passes).

WSS_z is the most likely beneficiary: on hood/roof surfaces where n̂ ≈ (0,0,1), the model's tau_z prediction conflates true tangential-z shear with normal-z force that should be zero. Current SOTA test_WSS_z = 8.6635% — the largest remaining sub-channel gap.

The surface normals are already in the batch: `surface_x[..., 3:6]` = `[nx, ny, nz]` (confirmed from `eval_tta_h252.py:211` where `mirror_inputs` negates `surface_x[..., 4]`).

### Step 1 — Add `--enforce-wss-tangential` flag to `eval_tta_h252.py`

Add a CLI argument:
```python
parser.add_argument("--enforce-wss-tangential", action="store_true", default=False,
                    help="Project WSS predictions onto surface tangent plane post-aggregation.")
```

### Step 2 — Apply projection after TTA aggregation

In `eval_tta_h252.py`, after the per-case TTA loop produces the aggregated surface prediction tensor `agg_surface_pred` (shape `[B, N, C]` where channels are `[cp, tau_x, tau_y, tau_z, ...]`), add:

```python
if args.enforce_wss_tangential:
    import torch.nn.functional as F
    # surface_x channels: [x, y, z, nx, ny, nz, area, ...]
    # Obtain normals from the (possibly mirrored) ORIGINAL batch — use non-mirrored surface_x
    n_hat = F.normalize(batch.surface_x[..., 3:6].float(), dim=-1)  # [B, N, 3]
    wss = agg_surface_pred[..., 1:4]   # [tau_x, tau_y, tau_z]
    normal_comp = (wss * n_hat).sum(-1, keepdim=True)  # [B, N, 1]
    wss_corrected = wss - normal_comp * n_hat
    agg_surface_pred = torch.cat([
        agg_surface_pred[..., :1],      # cp
        wss_corrected,                  # tau_x, tau_y, tau_z (corrected)
        agg_surface_pred[..., 4:],      # any remaining channels
    ], dim=-1)
```

**Important**: use the non-mirrored batch normals (the original batch before any y-flip). The mirror TTA already un-mirrors tau_y in `unmirror_surface_real` — by the time aggregation is done, the surface_x used for normals should be the original orientation.

### Step 3 — Diagnostic logging

Before applying the correction, log the mean normal-component magnitude so we know how much BC violation the model was making:
```python
if args.enforce_wss_tangential:
    mean_bc_violation = normal_comp.abs().mean().item()
    wss_magnitude = wss.norm(dim=-1).mean().item()
    print(f"BC violation: mean |τ·n̂| = {mean_bc_violation:.6f}, mean |τ| = {wss_magnitude:.6f}, ratio = {mean_bc_violation/wss_magnitude:.4%}")
```

Log this to W&B as `bc_violation_ratio`.

### Step 4 — Run on H296 config (K=4, 8-res, mirror)

```bash
H185_EP15_CKPT="outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt"

torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint $H185_EP15_CKPT \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-passes 4 \
  --antithetic-noise \
  --enforce-wss-tangential \
  --batch-size 2 --num-workers 4 \
  --agent thorfinn \
  --wandb-group "h305-thorfinn-bc-wss-normal-zero" \
  --wandb-name "thorfinn/h305-bc-K4-8res-mirror"
```

### Decision gate

- **val_abupt < 5.9221%**: continue to test arm
- **val_abupt ≥ 5.9221%**: BC projection is neutral/counterproductive — report `bc_violation_ratio` and close immediately (if ratio < 0.1%, the model's BC violation is negligible and this is a Finding)

## Baseline (H296 NEW SOTA — PR #1484)

```
val_abupt=5.9221%   test_abupt=5.7678%
test_VP=3.3763%     test_SP=3.6436%
test_WSS=6.6728%    test_WSS_z=8.6635%   ← primary target
W&B run: at1jadnv (fern/h296-ep15-anti-K4-8res-stack)
```

New merge gate: **val_abupt < 5.9221% AND test_abupt < 5.7678%**

Reproduce H296:
```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 4 --antithetic-noise \
  --batch-size 2 --num-workers 4
```

## Expected outcome

If BC violation ratio > 1%: likely WSS improvement, possibly val_abupt improvement.
If BC violation ratio < 0.1%: model already approximately satisfies BC, Finding "bc-null".
The key diagnostic is the violation ratio — please report it regardless of the metric result.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR (19:15Z): **Status check — H305 thorfinn, please respond.**

Your PR has been open since assignment with zero comments. This is the only experiment slot you have — we need this running now.

**H305 hypothesis**: Test whether reducing surface resolution ladder steps (fewer points but broader range) improves WSS prediction. The current SOTA uses 8 resolutions from 32768–131072. H305 proposes a sparser 6-step ladder with wider range.

Please confirm:
1. Have you read the assignment PR body and understood the experiment?
2. Are you currently running anything? If so, post your W&B run ID.
3. If you are blocked, post what is blocking you.

**Gate for this PR**: val_abupt < 5.9221% to proceed to test arm.

Please respond within 30 minutes or I will reassign this slot.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

Sorry for the silence — picking this up now. To clarify: the PR body describes H305 as **physics BC enforcement** (project WSS onto surface tangent plane post-TTA so τ·n̂ = 0), not a resolution-ladder change. I'll proceed with what's in the PR body.

**Current state:**
1. Implementation is in place on the branch (`eval_tta_h252.py` diff already on disk): `--enforce-wss-tangential` flag (auto-added via the dataclass-driven argparser), per-case projection of `agg_surface_pred[..., 1:4]` onto the tangent plane using non-mirrored normals from `case.surface_x[..., 3:6]`, plus diagnostic accumulators for `bc_violation_mean_normal_abs`, `bc_violation_mean_wss_norm`, and `bc_violation_ratio` (logged to W&B and printed at the end of each split).
2. About to smoke-test the code path with a tiny run, then kick off the full H296 config (K=4, 8-res, mirror, sigma=5e-4, antithetic) on the H185 EP15 checkpoint as specified in the PR.

Will post the W&B run ID as soon as the full eval launches.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

Update: an H305 BC eval was already running on this branch (must have been kicked off in a prior session). It started **2026-05-30 17:45 UTC** with the exact command from the PR body.

- **W&B run:** `evc7lnkf` — https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/evc7lnkf
- **Command:** `torchrun --standalone --nproc-per-node=8 eval_tta_h252.py --checkpoint outputs/h252_eval/_artifacts/0gjfv45i/epoch-15/checkpoint.pt --run-id 0gjfv45i --checkpoint-alias epoch-15 --resolutions 32768,40960,49152,57344,65536,81920,98304,131072 --eval-modes weight_noise_mirror_res_avg --weight-noise-sigma 5e-4 --weight-noise-passes 4 --antithetic-noise --enforce-wss-tangential --batch-size 2 --num-workers 4 --agent thorfinn --wandb-group h305-thorfinn-bc-wss-normal-zero --wandb-name thorfinn/h305-bc-K4-8res-mirror`
- **Status (19:33 UTC):** all 8 GPUs at 80-100% utilization, val split in-progress (2 of ~4-5 cases per rank done at ~40 min/case). ETA val+test completion ~01:30 UTC (≈6h from start). Within the 6h timeout but tight.

Will post results + the bc_violation_ratio diagnostic as soon as the run finishes.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["evc7lnkf"],"primary_metric":{"name":"full_val_primary/weight_noise_mirror_res_avg/abupt_axis_mean_rel_l2_pct","value":10.8591},"test_metric":{"name":"full_val_primary/weight_noise_mirror_res_avg/wall_shear_z_rel_l2_pct","value":21.8342}}

## Results

**Decision gate FAILED.** val_abupt = 10.8591% ≫ 5.9221% baseline. Test arm was killed per advisor instructions ("close immediately if val_abupt ≥ 5.9221%").

### Headline numbers

| Metric (val) | Baseline (H296 at1jadnv) | H305 (BC-enforced, evc7lnkf) | Δ |
|---|---|---|---|
| `abupt_axis_mean_rel_l2_pct` | 5.9221 | **10.8591** | **+4.94pp ❌** |
| `wall_shear_rel_l2_pct` | 6.7173 | **13.0828** | +6.37pp |
| `wall_shear_x_rel_l2_pct` | 5.8809 | 8.8844 | +3.00pp |
| `wall_shear_y_rel_l2_pct` | 7.2699 | 16.2041 | +8.93pp |
| `wall_shear_z_rel_l2_pct` (← target) | 9.0869 | **21.8342** | **+12.75pp ❌** |
| `surface_pressure_rel_l2_pct` | 3.9126 | 3.9126 | 0.00 (untouched) |
| `volume_pressure_rel_l2_pct` | 3.4603 | 3.4603 | 0.00 (untouched) |

### BC violation diagnostic

`bc_violation_ratio = 5.6412%` — i.e. mean |τ·n̂| / mean |τ| in the model's pre-correction predictions.

### What happened — root cause analysis

The hypothesis assumed targets satisfy τ·n̂ = 0 exactly because "the CFD solver enforces no-penetration at every surface point". I verified this assumption directly against the dataset (`run_4`, `run_22`, `run_56`, `run_109`, `run_150`, val split, all surface points per case):

| Case | mean \|τ_target · n̂\| | mean \|τ_target\| | ratio | mean ‖n‖-1 |
|---|---|---|---|---|
| run_4   | 0.112229 | 1.926244 | 5.83% | 0.000000 |
| run_22  | 0.107137 | 1.863843 | 5.75% | 0.000000 |
| run_56  | 0.117243 | 2.026404 | 5.79% | 0.000000 |
| run_109 | 0.093162 | 1.660117 | 5.61% | 0.000000 |
| run_150 | 0.106953 | 1.889138 | 5.66% | 0.000000 |
| **Total** | **0.106808** | **1.864939** | **5.7272%** | 0.000000 |

The processed DrivAerML dataset's *ground-truth* WSS vectors have a **~5.73% normal component** with respect to the stored surface normals (`surface_x[..., 3:6]`, which are unit length to machine precision). The model's predictions reproduce this artifact almost exactly (5.64% violation). Projecting model outputs onto the strict tangent plane therefore moves us systematically *away* from the targets by ~5.7% in the normal direction.

This dataset artifact is most likely a convention mismatch between the stored vertex normals and the face-level wall-shear sampling used during DrivAerML preprocessing — e.g. smoothed/area-weighted vertex normals not orthogonal to the local face element where τ_wall was integrated. The model has learned the convention faithfully; "physics enforcement" against a different normal convention destroys it.

### Hypothesis verdict

**Falsified.** The premise "CFD solver enforces τ·n̂ = 0 exactly" does not hold for the stored normals in this preprocessed dataset. BC projection is counterproductive, not zero-cost.

Note that *cp* (surface pressure) and volume pressure were untouched, as expected — those `rel_l2_pct` values match the H296 baseline to 4 decimal places, which confirms the eval setup is identical and isolates the regression to the WSS projection.

### Exact command run

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/h252_eval/_artifacts/0gjfv45i/epoch-15/checkpoint.pt \
  --run-id 0gjfv45i --checkpoint-alias epoch-15 \
  --resolutions 32768,40960,49152,57344,65536,81920,98304,131072 \
  --eval-modes weight_noise_mirror_res_avg \
  --weight-noise-sigma 5e-4 --weight-noise-passes 4 --antithetic-noise \
  --enforce-wss-tangential \
  --batch-size 2 --num-workers 4 \
  --agent thorfinn \
  --wandb-group h305-thorfinn-bc-wss-normal-zero \
  --wandb-name thorfinn/h305-bc-K4-8res-mirror
```

- **W&B run:** [evc7lnkf](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/evc7lnkf)
- **Peak GPU memory:** ~30.6 GB / GPU on 8× RTX PRO 6000 Blackwell (val split, 8-res ladder × K=4 antithetic noise × mirror)
- **Wall time:** val phase 11104 s (3h05m); test arm killed at start (~11s in) to save ~3 h of compute.

### Suggested follow-ups

1. **Recompute dataset normals to match the WSS sampling convention.** The 5.7% systematic normal component in targets strongly suggests the stored normals (`surface_x[..., 3:6]`) are not the same normals used during WSS extraction. If we can derive face-level / WSS-sampling normals from the raw mesh and store them as additional channels, BC projection becomes legitimate again. This is the only physically clean fix.

2. **Soft BC penalty during training, not hard projection at inference.** Add a regularizer `λ · ‖τ̂·n̂‖²` to the training loss with very small λ. The model can then learn how much normal component each target really has rather than us imposing it. Avoids the convention mismatch entirely.

3. **Estimate "true" τ on-the-fly via finite difference of pressure / SDF gradient.** If we want strict physics, compute the surface normal from the velocity gradient (∇u + ∇uᵀ) of an implicit field representation, then project. Out of scope for inference-only PRs.

4. **Re-verify on a single hood/roof point cluster.** I'd expect the dataset's normal-component ratio to be similar there to the global 5.7%. If it's larger near boundary/edge regions, that points more strongly to per-vertex-vs-per-face normal mismatch. (Not pursuing unless asked.)

5. **Try projecting in the opposite direction** — i.e. add `(τ̂·n̂)·n̂` instead of subtract, to test whether the sign of the dataset normal disagrees with the convention. (Cheap to run, would be diagnostic-only; I'd kill it before the test arm.)

### Files changed

- `eval_tta_h252.py`: added `--enforce-wss-tangential` flag, per-case post-aggregation projection of `agg_surface_pred[..., 1:4]` onto the tangent plane using `case.surface_x[..., 3:6]` normals, BC-violation diagnostic accumulators logged to W&B.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1496 because ADVISOR (20:55Z): Closing.

**Finding 'bc-tangent-projection-destructive'** banked: Post-TTA projection of WSS onto the surface tangent plane (zero normal component τ·n̂=0) gives val_abupt=10.8591% — a **+490bp catastrophic regression** vs H296 baseline 5.9221. Even though physically τ·n̂ must be zero at the wall, applying this projection to learned surface predictions destroys far more useful signal than the constraint-violation it removes. The model's surface normals (from `surface_x[..., 3:6]`) are noisy, so projecting onto them subtracts a substantial tangential signal.

Combined with the existing Finding 'coord-noise-harmful' (input geometry perturbation destroys predictions), this fully closes the **geometry-aware post-hoc correction axis** — the surface normals embedded in the dataset are not high-enough fidelity for hard physical constraints to be useful as TTA priors.

Test arm correctly killed at gate. Nothing to merge. Closing per dead-end protocol.

PR also has merge conflicts with tay; no rebase needed since this is being closed.

Good attempt, Thorfinn — pivoting you to a fresh axis..
